Add cache actuator endpoint for listing and clearing caches

joshiste opened the pull request:
This commits adds an actuator endpoint which lists the caches per
context and cacheManager and provides a delete operation to clear the
caches. As the statistics are exposed via the metrics endpoint they are
not included

closes #2625

<!--
Thanks for contributing to Spring Boot. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).
-->

philwebb commented:
Nice, thanks.

snicoll commented:
See #2625

snicoll commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cache/CacheEndpoint.java:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	@ReadOperation
+	public ApplicationCacheManagerBeans cacheManagerBeans() {
```
`cacheManagers` sounds more sensible to me.

snicoll commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cache/CacheEndpoint.java:
```diff
@@ -0,0 +1,156 @@
+	}
+
+	@DeleteOperation
+	public void clearCaches(@Nullable String contextId, @Nullable String cacheManagerName,
```
I wonder if we need to bring the complexity of `BeansEndpoint` here. If you are working with an application, I don't see the use case of having two cache managers with the same name in different context. 

As a user, having to specify the context is very annoying isn't it?

snicoll commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cache/CacheEndpoint.java:
```diff
@@ -0,0 +1,156 @@
+		}
+	}
+
+	private void clearCaches(ApplicationContext context, String cacheManagerName,
```
Naturally I would go with clearing a cache by name and just throw an exception if there are more than one `CacheManager` defining a cache with that name. Again, I don't think we should bring that complexity to the user for the very few that chose a complicated setup. The name of the cache manager could be used as a selector perhaps?

snicoll commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cache/CacheEndpoint.java:
```diff
@@ -0,0 +1,156 @@
+	 * Description of an application's {@link CacheManager} beans, primarily intended for
+	 * serialization to JSON.
+	 */
+	public static final class ApplicationCacheManagerBeans {
```
I have a feeling those names comes from the `BeansEndpoint`. I don't think this applies here.

joshiste commented:
@snicoll Thanks for your comments. I've copied some stuff from the liquibaseEndpoint hence the handling of multiple contexts.

joshiste commented:
I've renamed the endpoint to `CachesEndpoint`.
I've removed the handling of complex contexts.
I've removed the superfluous config metadata.

I didn't make the cacheManager a `@Selector` since having it as queryParam allows to delete all caches over all cacheManagers.

ptahchiev commented on spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cache/CachesEndpoint.java:
```diff
@@ -0,0 +1,109 @@
+			cacheManager.getCacheNames().forEach(cn -> cacheManager.getCache(cn).clear());
+		}
+		else {
+			cacheManager.getCache(cacheName).clear();
```
I believe  a null-pointer check is necessary here.

ptahchiev commented:
Maybe it's a good idea to add also a method for removing a single entry from the `Cache` region? Would be nice to also list all the entries in a given cache region, but the current spring cache API does not allow it (jcache does it allow it).

joshiste commented:
Added null check and rebased to master

joshiste commented:
>  I don't like the idea of using a bean name there either.
If we need to qualify the caches in case of duplicates names in two CacheManagers, I guess there is nothing else left to distinct the caches

How about rendering links for cache eviction into the response and just prefixing the cache name in case of ambiguity?
The downside is: it would need a mvc extension for the endpoint and doesn't clarify the usage in case of using the MXBean

```json
{
  "caches": [
    {
      "name": "manager-foo/ambiguousName",
      "evict": "/actuator/caches/manager-foo/ambiguousName"
    },
    {
      "name": "manager-bar/ambiguousName",
      "evict": "/actuator/caches/manager-bar/ambiguousName"
    },
    {
      "name": "uniqueName",
      "evict": "/actuator/caches/manager-bar/uniqueName"
    }
  ]
}
```
... if a cache contains a `/` it need's to be encoded...

snicoll commented:
@joshiste what I had in mind was more an attribute that would qualify the cache

```json
{
  "caches": [
    {
      "name": "ambiguousName",
      "cacheManager": "manager-foo",
      "_links": {
        "evict": "/actuator/caches/ambiguousName?cacheManager=manager-foo"
      }
    },
    {
      "name": "ambiguousName",
      "cacheManager": "manager-foo",
      "_links": {
        "evict": "/actuator/caches/ambiguousName?cacheManager=manager-bar"
      }
    },
    {
      "name": "uniqueName",
      "cacheManager": "manager-foo",
      "_links": {
        "evict": "/actuator/caches/manager-bar/uniqueName"
      }
    }
  ]
}
```

joshiste commented:
I've modified the output of the read operation as you suggested.

snicoll commented:
@joshiste thanks for the PR. While reviewing, I've reworked and significantly changed the initial proposal, see fb8a5a9 for more details.

joshiste commented:
Yay! Thanks for merging!
